### PR TITLE
C++ complex class fixes and tests

### PR DESCRIPTION
**Pull request by raffienficiaud**
* removing overload operator+ that seems to create a compilation issue
* adding unary operators + and -
* fixing type (polar)
* adding unit tests

Let me know if there is a particular linting/formatting to perform.

### CODE REVIEW

File: Cython/Includes/libcpp/complex.pxd [hunk 1]
```diff
@@ -8,21 +8,28 @@ cdef extern from "<complex>" namespace "std" nogil:
         # How to make the converting constructor, i.e. convert complex[double]
         # to complex[float]?
 
-        complex[T] operator+(complex[T]&)
-        complex[T] operator-(complex[T]&)
+        #complex[T] operator+(complex[T]&)
+        #complex[T] operator-(complex[T]&)
```
Comment (raffienficiaud) [inline Cython/Includes/libcpp/complex.pxd:12]: Left here for comments: those lines are preventing the `cythonization` of valid code

File: docs/examples/userguide/wrapping_CPlusPlus/python_to_cpp.pyx [hunk 2]
```diff
@@ -1,5 +1,7 @@
-# distutils: language = c++
+# cython: language_level=3
+# distutils: language=c++
```
Comment (raffienficiaud) [inline docs/examples/userguide/wrapping_CPlusPlus/python_to_cpp.pyx:2]: not sure if needed

File: docs/examples/userguide/wrapping_CPlusPlus/python_to_cpp.pyx [hunk 3]
```diff
@@ -17,3 +19,12 @@ print(vect)  # [1, 3, 5, 7, 9]
 
 cdef vector[string] cpp_strings = b'It is a good shrubbery'.split()
 print(cpp_strings[1])   # b'is'
+
+# creates a python object, then convert it to C++ complex
+complex_val = 1+2j
+cdef complex[double] c_value1 = complex_val
```
Comment (raffienficiaud) [inline docs/examples/userguide/wrapping_CPlusPlus/python_to_cpp.pyx:25]: I can't use 

```
cdef complex[double] c_value1 = 1+2j
```

directly with the error

```
python_to_cpp.pyx:24:33: Cannot assign type 'double complex' to 'complex[double]'
```

the generated `.cpp` code has mentions of `complex double` but those only appear in commented code.

File: docs/src/userguide/wrapping_CPlusPlus.rst [hunk 4]
```diff
@@ -607,9 +607,9 @@ with an underlying type::
 
     cdef extern from "Foo.h":
         cdef enum class Spam(unsigned int):
-	    x = 10
-	    y = 20
-	    ...
+            x = 10
+            y = 20
+            ...
```
Comment (raffienficiaud) [inline docs/src/userguide/wrapping_CPlusPlus.rst:612]: tabs vs space

**Comment by raffienficiaud** [comment]
I looked at the errors, one at least looks like this:

```
File "/home/runner/work/cython/cython/TEST_TMP/3/run/cpp/complex_numbers_cpp_directives_cpp_locals_True/complex_numbers_cpp.cpython-37m-x86_64-linux-gnu.so", line ?, in complex_numbers_cpp.scalar_double_conjugate
Failed example:
    a.imag
Expected:
    0.0
Got:
    -0.0
```

and it seems compiler dependent. I can make a quick fix for those.

**Comment by raffienficiaud** [comment]
I updated the failing test and rebased on latest master, please re-run.

File: Cython/Includes/libcpp/complex.pxd (hunk 1 above)
Comment (matusvalo) [reply Cython/Includes/libcpp/complex.pxd:12]: I think you can remove these lines

Comment (raffienficiaud) [reply Cython/Includes/libcpp/complex.pxd]: Done!

**Comment by raffienficiaud** [comment]
Added one commit for a spelling mistake from ac3e29771

**Comment by raffienficiaud** [comment]
Added tests on `cos`/`sin` and in-place operators `+`,`-`,`*`,`/` , minor fix on test.

**Comment by raffienficiaud** [comment]
Rebased onto latest master. Anything blocking the merge? Anything I can actually do?

**Comment by matusvalo** [comment]
> Rebased onto latest master. Anything blocking the merge? Anything I can actually do?

Thank you for your PR. If CI will be green I am OK with your PR but C++ is not my domain so I would like to have opinion from someone who is more familiar with C++.

**Comment by raffienficiaud** [comment]
> > Rebased onto latest master. Anything blocking the merge? Anything I can actually do?
> 
> Thank you for your PR. If CI will be green I am OK with your PR but C++ is not my domain so I would like to have opinion from someone who is more familiar with C++.

The CI was green before (now failing because of network issues `Service Unavailable [IP: 185.125.190.80 80]`), so I hope it stays green :)
Who can be a person more familiar with C++?

**Comment by matusvalo** [comment]
> Who can be a person more familiar with C++?

I think everyone except me :-D. @da-woods or @scoder but I am not sure how busy they are so we need to be patient.

**Comment by da-woods** [comment]
The PR looks fine to me. I'd like to make sure it's green first, and it looks like something external is failing still.

I'll try rerunning it tomorrow and merge it if it works then.

**Comment by raffienficiaud** [comment]
Thanks @matusvalo and @da-woods ! It was very green before the rebase and the errors seem like spurious network issues: rerunning should do.

**Comment by da-woods** [comment]
Thanks.

3.0.x commit is c73eb9ee0

**Comment by scoder** [comment]
Looks like the backport to 3.0.x does not work in Py2.7 for some reason:
```
======================================================================
FAIL: double_proj (line 233) (complex_numbers_cpp.__test__)
Doctest: complex_numbers_cpp.__test__.double_proj (line 233)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/doctest.py", line 2224, in runTest
    raise self.failureException(self.format_failure(new.getvalue()))
AssertionError: Failed doctest test for complex_numbers_cpp.__test__.double_proj (line 233)
  File "/home/runner/work/cython/cython/TEST_TMP/1/run/cpp/complex_numbers_cpp/complex_numbers_cpp.so", line unknown line number, in double_proj (line 233)

----------------------------------------------------------------------
File "/home/runner/work/cython/cython/TEST_TMP/1/run/cpp/complex_numbers_cpp/complex_numbers_cpp.so", line ?, in complex_numbers_cpp.__test__.double_proj (line 233)
Failed example:
    double_proj(-float("infinity") + 4j)
Expected:
    (-inf+0j)
Got:
    (inf+0j)
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/TEST_TMP/1/run/cpp/complex_numbers_cpp/complex_numbers_cpp.so", line ?, in complex_numbers_cpp.__test__.double_proj (line 233)
Failed example:
    double_proj(5 - float("infinity")*1j)
Expected:
    (inf+0j)
Got:
    (inf-0j)


======================================================================
FAIL: double_proj (line 233) (complex_numbers_cpp.__test__)
Doctest: complex_numbers_cpp.__test__.double_proj (line 233)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/doctest.py", line 2224, in runTest
    raise self.failureException(self.format_failure(new.getvalue()))
AssertionError: Failed doctest test for complex_numbers_cpp.__test__.double_proj (line 233)
  File "/home/runner/work/cython/cython/TEST_TMP/1/run/cpp/complex_numbers_cpp_directives_cpp_locals_True/complex_numbers_cpp.so", line unknown line number, in double_proj (line 233)

----------------------------------------------------------------------
File "/home/runner/work/cython/cython/TEST_TMP/1/run/cpp/complex_numbers_cpp_directives_cpp_locals_True/complex_numbers_cpp.so", line ?, in complex_numbers_cpp.__test__.double_proj (line 233)
Failed example:
    double_proj(-float("infinity") + 4j)
Expected:
    (-inf+0j)
Got:
    (inf+0j)
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/TEST_TMP/1/run/cpp/complex_numbers_cpp_directives_cpp_locals_True/complex_numbers_cpp.so", line ?, in complex_numbers_cpp.__test__.double_proj (line 233)
Failed example:
    double_proj(5 - float("infinity")*1j)
Expected:
    (inf+0j)
Got:
    (inf-0j)
```
https://github.com/cython/cython/actions/runs/8370946075/job/22919164368

**Comment by raffienficiaud** [comment]
@scoder actually the values returned by this test seem correct to me, according to the definition of [`std::proj(std::complex)`](https://en.cppreference.com/w/cpp/numeric/complex/proj). I'll look into this today.

**Comment by da-woods** [comment]
Probably https://github.com/cython/cython/pull/6111 - see if it works. I think the issue was that two `double_proj` functions were defined.

I've looked at the generated code and all it's doing is calling the standard library. What we want to test is that we call the right standard library functions, not that the standard library works. So if the standard library is being inconsistent we just need to make the definitions looser. But hopefully #6111 will sort it.

**Comment by raffienficiaud** [comment]
@da-woods good catch and thank you for fixing this!

**Pull request merged**